log/DESIGN.md

```diff
@@ -0,0 +1,301 @@
+for reducing the number of allocations when passing attributes.
+
+The bridge implementation should do its best to pass
+the `ctx` containing the trace context from the caller
+so it can later passed via `Emit`.
+Re-constructing a `context.Context` with [`trace.ContextWithSpanContext`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#ContextWithSpanContext)
+and [`trace.NewSpanContext`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#NewSpanContext)
+would usually involve more memory allocations.
+
+The logging libraries which have recording methods that accepts `context.Context`,
+such us [`slog`](https://pkg.go.dev/log/slog),
+[`logrus`](https://pkg.go.dev/github.com/sirupsen/logrus)
+[`zerolog`](https://pkg.go.dev/github.com/rs/zerolog),
+makes passing the trace context trivial.
+
+However, some libraries do not accept a `context.Context` in their recording methods.
+Structured logging libraries,
+such as [`logr`](https://pkg.go.dev/github.com/go-logr/logr)
+and [`zap`](https://pkg.go.dev/go.uber.org/zap),
+offer passing `any` type as a log attribute/field.
+Therefore, their bridge implementations can define a "special" log attributes/field
+that will be used to capture the trace context.
+
+[The prototype](https://github.com/open-telemetry/opentelemetry-go/pull/4725)
+has a naive implementation of
+[slog.Handler](https://pkg.go.dev/log/slog#Handler) in `log/internal/slog.go`
+and [logr.LogSink](https://pkg.go.dev/github.com/go-logr/logr#LogSink) in `log/internal/logr.go`.
+
+### Direct API usage
+
+The users may also chose to use the API directly.
+
+```go
+package app
+
+var logger = otel.Logger("my-service")
+
+// In some function:
+logger.Emit(ctx, Record{Severity: log.SeverityInfo, Body: "Application started."})
+```
+
+### API implementation
+
+If the implementation processes the record asynchronously,
+then it has to copy record attributes,
+in order to avoid use after free bugs and race condition.
+
+Excerpt of how SDK can implement the `Logger` interface.
+
+```go
+type Logger struct {
+	scope     instrumentation.Scope
+	processor Processor
+}
+
+func (l *Logger) Emit(ctx context.Context, r log.Record) {
+	// Create log record model.
+	record, err := toModel(r)
+	if err != nil {
+		otel.Handle(err)
+		return
+	}
+	l.processor.Process(ctx, record) // Note: A batch processor copies the attributes.
+}
+```
+
+A test implementation of the the `Logger` interface
+used for benchmarking is in `internal/writer_logger.go`.
+
+## Compatibility
+
+The backwards compatibility is achieved using the `embedded` design pattern
+that is already used in Trace API and Metrics API.
+
+Additionally, the `Logger.Emit` functionality can be extended by
+adding new exported fields to the `Record` struct.
+
+## Benchmarking
+
+The benchmarks take inspiration from [`slog`](https://pkg.go.dev/log/slog),
+because for the Go team it was also critical to create API that would be fast
+and interoperable with existing logging packages.[^1][^2]
```

Comment:
This does not seem relevant for an API design document.

Reply:
I disagree as benchmarks are validating the design from performance perpective.

Reply:
I think a benchmarking section is useful, but we should at least link to benchmarks we did that support our chosen approach.

Reply:
Do you think that adding a reference to https://github.com/open-telemetry/opentelemetry-go/blob/7fc957ab5a56c37b48a5223d9f01062ab8e74c32/log/internal/bench_test.go will be good enough or should do something more? I planned to update the design doc with hyperlinks to source files once they are in the repository. I updated the description of https://github.com/open-telemetry/opentelemetry-go/issues/3827 to not forget about it.

Reply:
IMO if there is a comment somewhere with the benchmark results that links to the benchmark code, that is even better.  But i'm OK with any reference that helps a future reader understand why we chose this approach.

Reply:
PTAL 31134025

Reply:
LGTM.